[SYNTH-12978] Make improvements to the README (#1372)
DataDog/datadog-ci: src/commands/synthetics/README.md

To setup the client, your Datadog API and application keys need to be configured. These keys can be defined in three different ways:

1. Defined in a [global JSON configuration file](#global-configuration-file-options):

    ```json
    {
      "apiKey": "<API_KEY>",
      "appKey": "<APPLICATION_KEY>",
    }
    ```

2. Defined as environment variables:

    ```bash
    export DATADOG_API_KEY="<API_KEY>"
    export DATADOG_APP_KEY="<APPLICATION_KEY>"
    ```

3. Passed to the CLI when running your tests:

    ```bash
    yarn datadog-ci synthetics run-tests --apiKey "<API_KEY>" --appKey "<APPLICATION_KEY>"
    ```

### Global configuration file options

Using a global configuration file is one of the ways to configure datadog-ci. To do so, create a JSON configuration file on your system. Specify the path to the file using the `--config` flag or configure it through the `DATADOG_SYNTHETICS_CONFIG_PATH` environment variable [when launching your tests](#run-tests-command) or [uploading a new application](#upload-application-command). If you don't specify a file path, Datadog looks for a file with the default filename of `datadog-ci.json`.

> See each command's list of configurations below for the list of advanced options in the global configuration file relevant to each [run-tests command](#list-of-configurations) and [upload-application command](#list-of-configurations-1). For an example configuration file, see this [`example-global-config.json` file][9].

Author: So originally we were using the `.github/workflows/e2e/global.config.json` and i was supposed to change that file to just include more info in [this PR](https://github.com/DataDog/datadog-ci/pull/1368), but then i realised that that file is actually used for something and is not just an example. So i opted to create a separate file just as an example instead.

As for if we should move it next to the README, considering that all of the other "config" files are in the `.github/workflows/e2e/` folder i think it makes more sense to leave the file in that folder. Additionally in terms of keeping this information close to the README the example just below is pretty much the same as this example file, so that should be good enough i think 🤔

> See each command's list of configurations below for the list of advanced options in the global configuration file relevant to each [run-tests command](#list-of-configurations) and [upload-application command](#list-of-configurations-1). For an example configuration file, see this [`example-global-config.json` file][9].

Author: @Drarig29 managed to convince me to move it :D 
Moved it to a examples folder in the synthetics folder 👍

  "tunnel": true
}
```

**Note**: The `global` field from the global configuration file is deprecated in favor of `defaultTestOverrides`.

### Environment Variables

In addition to the global configuration file, you can configure all properties using environment variables. If a property is defined in both the global configuration file and as an environment variable, the environment variable takes precedence.

Example:

```bash
export DATADOG_SITE=datadoghq.com
```

### Command line options

The CLI provides another way to set options and configure the behavior of datadog-ci. These options will override the global configuration file and environment variables.

Example:

```bash
yarn datadog-ci synthetics run-tests --public-id pub-lic-id1
```

The priority of the 3 forms of configuration is as follows:

```yml
Global Config < Environment variables < CLI parameters
```

### Using datadog-ci as a library

> You can also use the `datadog-ci` package as a library in your Node.js application. To do so, import the package and use the `execute()` function for either `runTests` or `uploadApplication`.

Reviewer: I tried to detail a bit this section, but couldn't get everything right. I think we should remove this section and cut another card to flesh this out with a proper example and instructions.

``````suggestion
You can also use the `datadog-ci` package as a library in your Node.js application to trigger tests. To do so, import the package from the Synthetics `run-tests` command and call the `executeWithDetails()` function.

``` javascript
import { executeWithDetails } from '@datadog/datadog-ci/dist/commands/synthetics/run-tests-lib';

const { results, summary } = await runTests.executeWithDetails(...)
```
``````

Reviewer 2: Why are you suggesting to import from `@datadog/datadog-ci/dist/commands/synthetics/run-tests-lib`?

We exported a [`synthetics` object](https://github.com/DataDog/datadog-ci/blob/fdf21c69bc79f35fe62b4356e881c3bdcfb29fd3/src/index.ts#L2) from the main index file for that purpose. The `executeWithDetails()` function was made for us internally because [we didn't want to break the current lib API](https://github.com/DataDog/datadog-ci/pull/1128#issuecomment-1838136065).

So I'd rather document the current "official" way to use the lib, or nothing (and refine the API as part of a card).

The "official" way is basically:
- Either the user only wants to call one "do-it-all" function, and they can do:
  ```ts
  import {synthetics} from '@datadog/datadog-ci'

  await synthetics.execute(
    {
      apiKey: '',
      appKey: '',
      publicIds: ['aaa-bbb-ccc'],
    },
    {}
  )
  ```
- Or they can use `synthetics.executeTests()` along with the exported utils to build their own logic (example taken [from our github action](https://github.com/DataDog/synthetics-ci-github-action/blob/fe4e6106a85c3439e2d90977cf4bda7f58183cd8/src/main.ts#L5-L41)):
  ```ts
  import {synthetics, utils} from '@datadog/datadog-ci'

  const config: synthetics.RunTestsCommandConfig = {
    // their config
  }

  const reporter = synthetics.utils.getReporter([new synthetics.DefaultReporter({context: process})])
  const startTime = Date.now()

  try {
    const {results, summary} = await synthetics.executeTests(reporter, config)
    const orgSettings = await synthetics.utils.getOrgSettings(reporter, config)

    synthetics.utils.renderResults({
      config,
      orgSettings,
      reporter,
      results,
      startTime,
      summary,
    })

    synthetics.utils.reportExitLogs(reporter, config, {results})

    const exitReason = synthetics.utils.getExitReason(config, {results})
    if (exitReason !== 'passed') {
      console.error(`Datadog Synthetics tests failed: ${printSummary(summary, config)}`)
    } else {
      console.info(`Datadog Synthetics tests succeeded: ${printSummary(summary, config)}`)
    }
  } catch (error) {
    synthetics.utils.reportExitLogs(reporter, config, {error})

    const exitReason = synthetics.utils.getExitReason(config, {error})
    if (exitReason !== 'passed') {
      console.error('Running Datadog Synthetics tests failed.')
    }
  }
  ```

    "host": "127.0.0.1",
    "port": 3128,
    "protocol": "http"
  }
}
```

> **Note**: The `global` field from the global configuration file is deprecated in favor of `defaultTestOverrides`.

Author: Yeah i think i started putting them initially just after the examples in which the filed's name was just changed as people that had been looking for an example that they had used before would be able to see it easily.

However now with the whole rearranging it doesn't make sense as much any more and i like your suggestion from one of the comments further down to pull it out in a specific section 👍

Configuration options:

* Global Config: N/A
* ENV variable: `DATADOG_SYNTHETICS_CONFIG_PATH=global-config.json`
* CLI param: `--config global-config.json`

#### `datadogSite`

The Datadog instance to which request is sent. The default is `datadoghq.com`.<!-- partial Your Datadog site is {{< region-param key="dd_site" code="true" >}}. partial -->

Configuration options:

* Global Config: `"datadogSite": "datadoghq.com"`
* ENV variable: `DATADOG_SITE=datadoghq.com`
* CLI param: `--datadogSite datadoghq.com`

#### `defaultTestOverrides`

Overrides for Synthetic tests applied to all tests.

Configuration options:

* Global Config: See `testOverrides` part of [Test files](#test-files) for an example.
* ENV variable: N/A
* CLI param: N/A

#### `failOnCriticalErrors`

A boolean flag that fails the CI job if no tests were triggered, or results could not be fetched from Datadog. The default is set to `false`.

Configuration options:

* Global Config: `"failOnCriticalErrors": true`
* ENV variable: `DATADOG_SYNTHETICS_FAIL_ON_CRITICAL_ERRORS=true`

> * CLI param: `--failOnCriticalErrors`

Author: I had completely forgotten that the `--no-...` was a thing :D 
Will add them 👍 

I actually can't even see in the code where it does that. Is that a default behaviour provided by clipanion i wonder 🤔

> * CLI param: `--failOnCriticalErrors`

Reviewer 2: Yes it is: https://mael.dev/clipanion/docs/options#booleans

Configuration options:

* Global Config: `"failOnMissingTests": true`
* ENV variable: `DATADOG_SYNTHETICS_FAIL_ON_MISSING_TESTS=true`
* CLI param: `--failOnMissingTests`

#### `failOnTimeout`

A boolean flag that fails the CI job if at least one test exceeds the default test timeout. The default is set to `true`.

Configuration options:

* Global Config: `"failOnTimeout": true`
* ENV variable: `DATADOG_SYNTHETICS_FAIL_ON_TIMEOUT=true`
* CLI param: `--failOnTimeout`

#### `files`

Glob patterns to detect Synthetic test [configuration files](#test-files).

Configuration options:

* Global Config: `"files": ["{,!(node_modules)/**/}*.synthetics.json"]`
* ENV variable: `DATADOG_SYNTHETICS_FILES="{,!(node_modules)/**/}*.synthetics.json"`
* CLI param: `-f "{,!(node_modules)/**/}*.synthetics.json"` / `--files "{,!(node_modules)/**/}*.synthetics.json"`

#### `jUnitReport`

The filename for a JUnit report if you want to generate one.

Configuration options:

> * Global Config: `"jUnitReport": "e2e-test-junit"`

Author: I just used the value used in the example in the reporter section
https://github.com/DataDog/datadog-ci/pull/1372/files#diff-5a0fda3a46d1c105f78a54a8d87902bdb9886b1739928f72abc7cefc23a2f86bR906
but i do like the idea of specifying that it will create an `.xml` file

* CLI param: `--selectiveRerun`

#### `subdomain`

The name of the custom subdomain set to access your Datadog application. If the URL used to access Datadog is `myorg.datadoghq.com`, the `subdomain` value needs to be set to `myorg`.

Configuration options:

* Global Config: `"subdomain": "myorg"`
* ENV variable: `DATADOG_SUBDOMAIN="myorg"`
* CLI param: `--subdomain "myorg"`

#### `testSearchQuery`

Pass a query to select which Synthetic tests to run. If you are running tests in the CLI, use the `-s` flag.

Configuration options:

* Global Config: `"testSearchQuery": "tag:e2e-tests"`
* ENV variable: `DATADOG_SYNTHETICS_TEST_SEARCH_QUERY="tag:e2e-tests"`
* CLI param: `-s "tag:e2e-tests"` / `--search "tag:e2e-tests"`

#### `tunnel`

Use [Local and Staging Environments](#use-local-and-staging-environments) to execute your test batch.

Configuration options:

* Global Config: `"tunnel": true`
* ENV variable: `DATADOG_SYNTHETICS_TUNNEL=true`
* CLI param: `-t` / `--tunnel`

**Note**: The `pollingTimeout` option and `--pollingTimeout` CLI parameter are deprecated in favor of `batchTimeout` and `--batchTimeout`, respectively.

### Failure modes flags

> * `--failOnTimeout` (or `--no-failOnTimeout`) makes the CI fail (or pass) if one of the results exceeds its test timeout.

Author: I didn't really consider the contents here tbf i just moved it, but yeah i just double check the information here against the descriptions and they're basically the same, so i think it does not make much sense to keep it.

> * `--failOnTimeout` (or `--no-failOnTimeout`) makes the CI fail (or pass) if one of the results exceeds its test timeout.

Reviewer 2: We could say:
> `--no-...` can be used for any boolean CLI parameter, e.g. `--no-failOnTimeout`

> * `--failOnTimeout` (or `--no-failOnTimeout`) makes the CI fail (or pass) if one of the results exceeds its test timeout.

Author: I just added them to the examples for all the cli parameters that are booleans 👍

Configuration options:

* Global Config: `"batchTimeout": 180000`
* ENV variable: `DATADOG_SYNTHETICS_BATCH_TIMEOUT=180000`
* CLI param: `--batchTimeout 180000`

#### `configPath`

The global JSON configuration is used when launching tests. See the [example configuration](https://docs.datadoghq.com/continuous_testing/cicd_integrations/configuration/?tab=npm#global-configuration-file-options ) for more details.

Configuration options:

* Global Config: N/A
* ENV variable: `DATADOG_SYNTHETICS_CONFIG_PATH=global-config.json`
* CLI param: `--config global-config.json`

#### `datadogSite`

The Datadog instance to which request is sent. The default is `datadoghq.com`.<!-- partial Your Datadog site is {{< region-param key="dd_site" code="true" >}}. partial -->

Configuration options:

* Global Config: `"datadogSite": "datadoghq.com"`
* ENV variable: `DATADOG_SITE=datadoghq.com`
* CLI param: `--datadogSite datadoghq.com`

#### `defaultTestOverrides`

Overrides for Synthetic tests applied to all tests.

Configuration options:

> * Global Config: See `testOverrides` part of [Test files](#test-files) for an example.

Reviewer 2: This shows that we could improve the structure of the README: I think we should have a section named "Test overrides", which would list the possible test overrides

And it would say that those overrides can either be applied to all tests with `defaultTestOverrides`, or to some specific tests with their `testOverrides`.

The "Test files" section would then be smaller.

Configuration options:

* Global Config: `"selectiveRerun": true,`
* ENV variable: `DATADOG_SYNTHETICS_SELECTIVE_RERUN=true`
* CLI param: `--selectiveRerun` / `--no-selectiveRerun`

#### `subdomain`

The name of the custom subdomain set to access your Datadog application. If the URL used to access Datadog is `myorg.datadoghq.com`, the `subdomain` value needs to be set to `myorg`.

Configuration options:

* Global Config: `"subdomain": "myorg"`
* ENV variable: `DATADOG_SUBDOMAIN="myorg"`
* CLI param: `--subdomain "myorg"`

#### `testSearchQuery`

Pass a query to select which Synthetic tests to run.

Configuration options:

* Global Config: `"testSearchQuery": "tag:e2e-tests"`
* ENV variable: `DATADOG_SYNTHETICS_TEST_SEARCH_QUERY="tag:e2e-tests"`
* CLI param: `-s "tag:e2e-tests"` / `--search "tag:e2e-tests"`

#### `tunnel`

Use [Local and Staging Environments](#use-local-and-staging-environments) to execute your test batch.

Configuration options:

* Global Config: `"tunnel": true`
* ENV variable: `DATADOG_SYNTHETICS_TUNNEL=true`
* CLI param: `-t` / `--tunnel` / `--no-tunnel`

> ### Configure a StartUrl

Reviewer 2: ```suggestion
### Configure a start URL
```

```bash
export DATADOG_SUBDOMAIN="myorg"
```

### Configure custom locations

You can use `DATADOG_SYNTHETICS_OVERRIDE_LOCATIONS` to override the locations where your tests run. Locations should be separated with `;`. The configuration in [test files](#test-files) takes precedence over other overrides.

```bash
export DATADOG_SYNTHETICS_OVERRIDE_LOCATIONS="aws:us-east-1;aws:us-east-2"
```

### Test files

Test configuration files let you customize individual tests or set up multiple runs of the same test with different settings, beyond what you can do with other configuration methods.

These files take precedence over global configuration files, environment variables, and CLI parameters. So, the priority order including test configurations is as follows:

``` yml
Global Config < Environment variables < CLI parameters < Test Config
```

By default, `datadog-ci` runs at the root of the working directory and looks for `{,!(node_modules)/**/}*.synthetics.json` files (every file ending with `.synthetics.json`, except for those in the `node_modules` folder) to find a test configuration file. This can be manually configured with the [`files` parameter](#files).

> Example Test configuration file:

Reviewer 2: ```suggestion
Example:
```

        "testTimeout": 300,
        "variables": {"MY_VARIABLE": "new title"}
      }
    },
    {
      "id": "<TEST_PUBLIC_ID_2>",
      "testOverrides": {
        "allowInsecureCertificates": true,
        ...
        "variables": {"MY_VARIABLE": "new title"}
      }
    }
  ]
}
```

**Note**: The `config` field from the test configuration file is deprecated in favor of `testOverrides`.

The `<TEST_PUBLIC_ID>` can be either the identifier of the test found in the URL of a test details page (for example, for `https://app.datadoghq.com/synthetics/details/abc-def-ghi`, it would be `abc-def-ghi`) or the full URL to the details page (for example, directly `https://app.datadoghq.com/synthetics/details/abc-def-ghi`).

All options under the `testOverrides` key are optional and allow overriding of the test configuration as stored in Datadog. These options can also be set with environment variables starting with `DATADOG_SYNTHETICS_OVERRIDE_`, or with the `--override` CLI parameter following this pattern: `--override option=value`.

#### `allowInsecureCertificates` (Boolean)

Disable certificate checks in Synthetic API and Browser tests.

Configuration options:

> * Test Config: `"allowInsecureCertificates": true`

Reviewer 2: The namings `Global Config` and `Test Config` (with those uppercase letters, and abbreviation) are only internal, it would be better to align and say the same thing everywhere

> * Test Config: `"allowInsecureCertificates": true`

Reviewer 2: My bad, I see you used it everywhere already. Could we make it explicit somewhere that `Global Config` is short for "global configuration file", and "Test Config" is short for "test (configuration) file"?

It's important to me that we use the same namings everywhere

        "testTimeout": 300,
        "variables": {"MY_VARIABLE": "new title"}
      }
    },
    {
      "id": "<TEST_PUBLIC_ID_2>",
      "testOverrides": {
        "allowInsecureCertificates": true,
        ...
        "variables": {"MY_VARIABLE": "new title"}
      }
    }
  ]
}
```

The `<TEST_PUBLIC_ID>` can be either the identifier of the test found in the URL of a test details page (for example, for `https://app.datadoghq.com/synthetics/details/abc-def-ghi`, it would be `abc-def-ghi`) or the full URL to the details page (for example, directly `https://app.datadoghq.com/synthetics/details/abc-def-ghi`).

> All options under the `testOverrides` key are optional and allow overriding of the test configuration as stored in Datadog. These options can also be set with environment variables starting with `DATADOG_SYNTHETICS_OVERRIDE_`, or with the `--override` CLI parameter following this pattern: `--override option=value`.

Reviewer 2: Could we also mention `defaultTestOverrides` here?

If the URL contains variables, this regex applies after the interpolation of the variables.

There are two possible formats:

* **`your_regex|your_substitution`**: The pipe-based syntax, to avoid any conflicts with / characters in URLs. For example, `https://example.com(.*)|http://subdomain.example.com$1` to transform `https://example.com/test` to `http://subdomain.example.com/test`.
* **`s/your_regex/your_substitution/modifiers`**: The slash syntax, which supports modifiers. For example, `s/(https://www.)(.*)/$1extra-$2/` to transform `https://www.example.com` into `https://www.extra-example.com`.

Configuration options:

* Global/Test Config: `"startUrlSubstitutionRegex": "s/(https://www.)(.*)/$1extra-$2/"`
* ENV variable: `DATADOG_SYNTHETICS_OVERRIDE_START_URL_SUBSTITUTION_REGEX="s/(https://www.)(.*)/$1extra-$2/"`
* CLI param: `--override startUrlSubstitutionRegex="s/(https://www.)(.*)/$1extra-$2/"`

#### `testTimeout` (Number)

The maximum duration of a browser test in seconds.

Configuration options:

* Global/Test Config: `"testTimeout": 300`
* ENV variable: `DATADOG_SYNTHETICS_OVERRIDE_TEST_TIMEOUT=300`
* CLI param: `--override testTimeout=300`

#### `variables` (Object)

> The variables to replace in the test. This object should contain key as the name of the variable to replace and values as the new value of the variable to replace.

Reviewer 3: ```suggestion
This object defines the variables to be substituted in the test. It should include keys corresponding to the names of the variables to be replaced, and values representing the new values for these variables.
```

I was a little confused by this sentence too so suggest to re-word for clarity.

* CLI param:
  * `--override variables.NEW_VARIABLE_1="NEW VARIABLE 1"`
  * `--override variables.NEW_VARIABLE_2="NEW VARIABLE 2"`

## Upload Application Command

This command uploads a new version to an **existing** mobile application.

### List of Configurations

<!--
  The descriptions of these is kept for consistency in this doc - https://datadoghq.atlassian.net/wiki/spaces/SYN/pages/3378446383/Configuration+parameters+and+their+usages#Proposal-for-aligning-Descriptions-and-labels
  If want to update them please update the doc and the relevant integrations as well.
-->

#### `apiKey`

The API key used to query the Datadog API.

Configuration options:

* Global Config: `"apiKey": "<API_KEY>"`
* ENV variable: `DATADOG_API_KEY="<API_KEY>"`
* CLI param: `--apiKey "<API_KEY>"`

#### `appKey`

The application key used to query the Datadog API.

Configuration options:

* Global Config: `"appKey": "<APPLICATION_KEY>"`
* ENV variable: `DATADOG_APP_KEY="<APPLICATION_KEY>"`
* CLI param: `--appKey "<APPLICATION_KEY>"`

#### `configPath`

> The global JSON configuration is used when launching tests. See the [example configuration](https://docs.datadoghq.com/continuous_testing/cicd_integrations/configuration/?tab=npm#global-configuration-file-options ) for more details.

Reviewer 3: ```suggestion
The global JSON configuration is used when launching tests. See the [example configuration](#global-configuration-file-options) for more details.
```

* `username` (String): The username for basic authentication.
* `password` (String): The password for basic authentication.

Configuration options:

* Test Config: `"basicAuth": {"username": "test_username", "password": "test_password"}`
* ENV variable:
  * `DATADOG_SYNTHETICS_OVERRIDE_BASIC_AUTH_USERNAME=test_username`
  * `DATADOG_SYNTHETICS_OVERRIDE_BASIC_AUTH_PASSWORD=test_password`
* CLI param:
  * `--override basicAuth.username=test_username`
  * `--override basicAuth.password=test_password`

#### `body` (String)

Data to send in an API test.

Configuration options:

* Test Config: `"body": "{\"fakeContent\":true}"`
* ENV variable: `DATADOG_SYNTHETICS_OVERRIDE_BODY={"fakeContent":true}`
* CLI param: `--override body={"fakeContent":true}`

#### `bodyType` (String)

Content type for the data to send in an API test.

Configuration options:

* Test Config: `"bodyType": "application/json"`
* ENV variable: `DATADOG_SYNTHETICS_OVERRIDE_BODY_TYPE=application/json`
* CLI param: `--override bodyType=application/json`

#### `cookies` (String or object)

Use the provided string as a cookie header in an API or browser test (in addition or as a replacement).

> * If this is a string, it is used to replace the original cookies as `append` defaults to `false`.

Author: Well this can be easily solved by just swapping the order of the 2 bullet points :D 
but i think that it does actually sound better as you suggest it and it doesn't need explain internal logic
